### ISSUE TITLE
Add "allow" rule to Windows firewall for WSL2 network

### ISSUE THREAD

**Issue by throwable-one** [microsoft/WSL#4585]
I was trying to connect to Windows machine from Linux and found that it is blocked by default.

WSL virtual NIC connection belongs to "Public" profile and almost all connections are forbidden. 

As workaround, I added following rule explicitly:
```
PS C:\WINDOWS\system32> $myIp = (Ubuntu1804 run "cat /etc/resolv.conf | grep nameserver | cut -d' ' -f2")
PS C:\WINDOWS\system32> $myIp
172.21.0.1
PS C:\WINDOWS\system32>  New-NetFirewallRule -DisplayName "WSL" -Direction Inbound  -LocalAddress $myIp -Action Allow
```

Workaround works. But it would be great to have this rule enabled by default.

**Comment by throwable-one** [comment]
How to repro.

On Windows, listen to some port
```
PS C:\WINDOWS\system32> $Listener = [System.Net.Sockets.TcpListener]80;
PS C:\WINDOWS\system32> $Listener.Start();
PS C:\WINDOWS\system32> $Listener.AcceptSocket();
```

On WSL2, try to connect to it:
```$ telnet $(cat /etc/resolv.conf | grep nameserver | cut -d' ' -f2) 80```

You would not be able to connect until you create appropriate firewall rule

**Comment by therealkenc** [comment]
#4139

**Comment by dansanduleac** [comment]
Nice!
Since the IP might change every time you restart your computer, why not use the inteface?

```powershell
New-NetFirewallRule -DisplayName "WSL" -Direction Inbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow
```

**Comment by eromoe** [comment]
@dansanduleac This allow ping, but how can I access some port on windows , without shutdown the firewall ?

**Comment by Dicridon** [comment]
> Nice!
> Since the IP might change every time you restart your computer, why not use the inteface?
> 
> ```powershell
> New-NetFirewallRule -DisplayName "WSL" -Direction Inbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow
> ```

Thank you so much, I searched for about two hours to make WSL2 launch a graphic Emacs, and this command enabled the use of Emacs GUI, many thanks!

**Comment by HQDragon** [comment]
> @dansanduleac This allow ping, but how can I access some port on windows , without shutdown the firewall ?

try add new inbound rule for these ports?

**Comment by dansanduleac** [comment]
@eromoe @HQDragon The command I posted should allow you to access anything exposed by Windows from WSL, no matter what port, however bear in mind that any apps you've launched get an automated rule created for them when you first launch them, blocking access from public networks (this is when you get a prompt from Windows Firewall, asking whether the app should be allowed to accept connections from public networks). If you _don't explicitly allow_, they will be blocked by default, which also blocks connections from WSL.  So you might need to find that inbound rule, and change it from block to allow (or just delete it).

More info [here](https://github.com/cascadium/wsl-windows-toolbar-launcher#firewall-rules) (linked from this comment on the original issue https://github.com/microsoft/WSL/issues/4139#issuecomment-607477784)

**Comment by chaoky** [comment]
> Nice!
> Since the IP might change every time you restart your computer, why not use the inteface?
> 
> ```powershell
> New-NetFirewallRule -DisplayName "WSL" -Direction Inbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow
> ```

this should in the docs

**Comment by yakaboskic** [comment]
**In case this helps anyone:** I was trying to setup a new laptop that I just bought and nothing here worked. That's when I realized the Windows machine I just bought had a McAfee Firewall on it (free one year subscription). So all the windows firewall stuff wasn't actually working cause it wasn't the active Firewall (McAfee was) and they have known [compatibility issues with WSL/WSL2](https://kc.mcafee.com/corporate/index?page=content&id=KB91411). Once I uninstalled the McAfee firewall (you could also just disable it), everything here worked except the telnet stuff. For some reason that still didn't work, but I was able to connect to my X servers no problem, which is probably why most of us were here.

**Comment by brucejo75** [comment]
If you are trying to connect to a `mongodb` follow @dansanduleac [advice](https://github.com/microsoft/WSL/issues/4585#issuecomment-643678552).

Mongo installs rules `MongoDB Database Server` on windows that `disallow` public connections to mongodb.  To be able to access from WSL these need to be disabled.

That is 6 hours I won't get back!  I hope you are saved that time.

**Comment by balexandre** [comment]
this is still extremely unreliable,
sometimes works, sometimes it does not, other times only works until you shut down the computer 😢

I've 
- restored all Firewall rules
- added the `WSL` rule as `WSL by interface` as 
`New-NetFirewallRule -DisplayName "WSL by interface" -Direction Inbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow`
- tested without restarting
- tested after restarting

I can't reach the `localhost:8989` from the Windows side and as Ubuntu-20.04 under WSL2 does not have any browser, it's hard to develop in these conditions 😔

![image](https://user-images.githubusercontent.com/45473/115764494-e90ff300-a3a5-11eb-9544-d6a7637256bc.png)

Does anyone know some reliable trick that I can use to develop web interfaces under WSL2?

**Comment by ahillikerMDSOL** [comment]
> Does anyone know some reliable trick that I can use to develop web interfaces under WSL2?

You can hit the IP address that Ubuntu is using if you're using the same machine (ifconfig from WSL2 terminal will give you this address) - is that what you mean?

What I'm running into is trying to hit that IP from a wifi connected device like a phone or something.

I'm definitely in over my head here, but was hoping changing the -InterfaceAlias to 'vWan' or something would do the trick, but a) I don't know where to find the correct term for that alias, and b) I don't know if that's even how it works - maybe WSL interprets incoming wifi requests through the vEthernet adapter?

**Comment by celowsky** [comment]
I'm still having this issue as well when trying to debug my application from within the WSL using Intellij's Pycharm

**Comment by stopal** [comment]
> @eromoe @HQDragon The command I posted should allow you to access anything exposed by Windows from WSL, no matter what port, however bear in mind that any apps you've launched get an automated rule created for them when you first launch them, blocking access from public networks (this is when you get a prompt from Windows Firewall, asking whether the app should be allowed to accept connections from public networks). If you _don't explicitly allow_, they will be blocked by default, which also blocks connections from WSL. So you might need to find that inbound rule, and change it from block to allow (or just delete it).
> 
> More info [here](https://github.com/cascadium/wsl-windows-toolbar-launcher#firewall-rules) (linked from this comment on the original issue [#4139 (comment)](https://github.com/microsoft/WSL/issues/4139#issuecomment-607477784))

Thank you so much for that hint! I was about to give up. Setting up the specific IP Range for the "Remote IP Adresses" has done the trick for me. -> https://github.com/microsoft/WSL/issues/4139#issuecomment-618019890

**Comment by DiegoPrestesGit** [comment]
Sometimes I have problems with the WSL network, but this always do the trick: ``Get-Service LxssManager | Restart-Service`` this just reboots the WSL, and I can access "localhost:4321" normally again, I have to run this every time I start Windows.

**Comment by davelet** [comment]
I resolved it as this thread: https://github.com/microsoft/WSL/issues/4139#issuecomment-1016226909

**Comment by mcicat** [comment]
Thank You! This solved an issue I've been facing for over a month. For some reason after updating to Windows 11, fetching repositories by VSCode (which was started from WSL) was unbearably slow. This solved it!

**Comment by marriagav** [comment]
> 

Does anyone know how to do this for an outbound rule? I´ve tried just changing the direction to no success

**Comment by MrPointer** [comment]
> > 
> 
> Does anyone know how to do this for an outbound rule? I´ve tried just changing the direction to no success

@marriagav Very naively, I tried this:  
`New-NetFirewallRule -DisplayName "WSL" -Direction Outbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow`

And it worked for me! This is the only thing that has worked in fact, among tons of others that I've tried 😄

**Comment by merajhasan88** [comment]
I tried all of the above as well as this thread: [https://github.com/microsoft/WSL/issues/4139] to no avail. Windows 21H2 

Just trying to use GTK-3 on VSCode with a WSL2 Ubuntu terminal. Using VcXsrv running in background. 
I run this command on a small Hello World code:`
`gcc `pkg-config --cflags gtk+-3.0` -o gui hello.c `pkg-config --libs gtk+-3.0``
and it gives a 'gui' object. I run:
`./gui`
and it says: ```

Unable to init server: Could not connect: Connection refused

(gui:32): Gtk-WARNING **: 10:51:06.383: cannot open display:
```

I installed xfce4 but that doesnt work too. Upon running this:
`xfce4-session`
I get this:```

Unable to init server: Could not connect: Connection refused
xfce4-session: Cannot open display: .
Type 'xfce4-session --help' for usage.
```

Restarted everything several times.

**Comment by kenvix** [comment]
For my instances, `New-NetFirewallRule` doesn't work and I don't know why. I use `Set-NetFirewallProfile` to directly disable firewall on WSL2 interface instead to add a rule and allow it.

```powershell
#Requires -RunAsAdministrator

Set-NetFirewallProfile -Profile Private -DisabledInterfaceAliases "vEthernet (WSL)"
Set-NetFirewallProfile -Profile Public -DisabledInterfaceAliases "vEthernet (WSL)"
```

**Comment by KeisukeeLee** [comment]
> > > 
> > 
> > 
> > Does anyone know how to do this for an outbound rule? I´ve tried just changing the direction to no success
> 
> @marriagav Very naively, I tried this: `New-NetFirewallRule -DisplayName "WSL" -Direction Outbound -InterfaceAlias "vEthernet (WSL)" -Action Allow`
> 
> And it worked for me! This is the only thing that has worked in fact, among tons of others that I've tried 😄

I tested this. This is work for not only WSL NIC but all Interfaces.

**Comment by schmik** [comment]
Thanks, everyone. WIN11 here, _Rancher Desktop_ with _containerd_.

PowerShell'ing

`Get-NetAdapter -IncludeHidden | Format-Table -AutoSize`

helped find the right "-InterfaceAlias"...

I ended up using

`New-NetFirewallRule -DisplayName "WSL" -Direction Outbound  -InterfaceAlias "vEthernet (WSL (Hyper-V firewall))"  -Action Allow`

Now, when I have a containerized pgadmin and a containerized postgres listening on :5432...

```
nerdctl ps
CONTAINER ID    IMAGE                                COMMAND                   CREATED           STATUS    PORTS                     NAMES
5d0c3acb38ea    docker.io/dpage/pgadmin4:latest      "/entrypoint.sh"          14 minutes ago    Up        0.0.0.0:6543->80/tcp      pgadmin4-5d0c3
92147e941b71    docker.io/library/postgres:latest    "docker-entrypoint.s…"    17 minutes ago    Up        0.0.0.0:5432->5432/tcp    localDB
```

... I'm (pgadmin is) able to connect to **host.docker.internal**:5432 👌

**Comment by LifeJustDLC** [comment]
Because Windows randomly turns on the firewall without your acknowledge, enforce a cron job to disable it might be the only compromised solution.
```powershell
# run as admin in "windows powershell"
$option = New-ScheduledJobOption -RunElevated -RequireNetwork
$atLogin = New-JobTrigger -AtLogOn -User *
$atInterval = New-JobTrigger -Once -RepeatIndefinitely -RepetitionInterval "00:10:00" -At "0am"
Register-ScheduledJob -Name "F wsl firewall" -ScheduledJobOption $option -Trigger $atLogin, $atInterval -ScriptBlock {
    Set-NetFirewallProfile -Profile Public -DisabledInterfaceAliases "vEthernet (WSL)";
}
```
(I know basically nothing about windows powershell, so please let me know if I'm wrong.)
Also don't forget to enable inbound rules on the "Windows" host side.

**Comment by pikouser** [comment]
For me command "New-NetFirewallRule -DisplayName "WSL" -Direction Outbound  -InterfaceAlias "vEthernet (WSL (Hyper-V firewall))"  -Action Allow" fails with HRESULT 0x80070057. Adding this rule manually also does not change ports blockage, seems block is inside WSL. Anyway I found other solution. Just storing it here so perhaps that solves someone's time. It is already possible to run GUI apps with WSL:
https://learn.microsoft.com/en-us/windows/wsl/tutorials/gui-apps

**Comment by dante198712** [comment]
I just install WSL Ubuntu in W11 and nothing works to do anywhere, it also happened to me the same thing testing on other virtualization platforms such as Virtualbox, Docker and Hyper-V.

The easiest thing was to completely destabilize the Windows Firewall, but it did not work, additionally I have already tried a lot of tutorials related to adding Firewal Gui/Powershell rules and still totally blocked.

Very bad that it is so complicated to enable such basic functionality, formerly even in Virtualbox it was very easy to use the internet from the guest system!.

**Comment by Petros626** [comment]
I did the following, which worked for me (Windows 10 & WSL2):
1. `sudo nano /etc/ssh/ssh_config`
2. Under section **Host*** add **ServerAliveInterval 5** (I guess any value is okay?)
3. `sudo nano /etc/ssh/sshd_config`
4. Near the end of the file search for: **ClientAliveInterval 60**; **TCPKeepAlive yes**; **ClientAliveCountMax 10000**
5. `service ssh restart`
6. `ssh nameofmachine@xxx.xxx.xxx.xx`

**Comment by dwymark-celestron** [comment]
These are the commands I needed to get the internet working again. Not sure which of these were necessary; posting this here for my own future reference. I'll update it if I learn more later.

```powershell
New-NetFirewallRule -DisplayName "WSL" -Direction Outbound -InterfaceAlias "vEthernet (WSL (Hyper-V firewall))" -Action Allow
New-NetFirewallRule -DisplayName "WSL" -Direction Outbound -InterfaceAlias "vSwitch (Default Switch)" -Action Allow
wsl --shutdown
```

Then I restarted WSL by opening a new session in Windows Terminal.

I used the command below to find the two InterfaceAliases (thanks @schmik):
```powershell
Get-NetAdapter -IncludeHidden | Format-Table -AutoSize
```

**Comment by dwymark-celestron** [comment]
UPDATE: I realized that there is another factor that was important in my case. For an unrelated reason, I needed to run the following commands in PowerShell:
```
net stop hns
net start hns
```

Doing this while WSL2 is running causes it to lose internet access even though DNS continues to function. I am able to get the internet access to work again by restarting WSL. (Specifically, I mean running `wsl --shutdown` and then opening Ubuntu in Windows Terminal.)

I am not sure if the firewall commands I mentioned in my last comment were important or not, because I also restarted WSL. It's possible that they were just a red herring (in my case).

**Comment by Zon1234** [comment]
for me, this fixed the issue
https://learn.microsoft.com/de-de/windows/security/operating-system-security/network-security/windows-firewall/hyper-v-firewall

**Comment by cloudqq** [comment]
> Nice! Since the IP might change every time you restart your computer, why not use the inteface?
> 
> New-NetFirewallRule -DisplayName "WSL" -Direction Inbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow


2025 Windows 11 24H2

`
 New-NetFirewallRule -DisplayName "WSL" -Direction Inbound  -InterfaceAlias "vEthernet (WSL (Hyper-V firewall))"  -Action Allow
`

**Comment by willser** [comment]
env:
``` 
systeminfo | findstr /B /C:"OS 名称" /C:"OS 版本"
OS 名称:          Microsoft Windows 10 专业版
OS 版本:          10.0.19045 暂缺 Build 19045
```

I use this command `New-NetFirewallRule -DisplayName "WSL by interface" -Direction Inbound  -InterfaceAlias "vEthernet (WSL)"  -Action Allow` after run wsl and it works,but after windows reboot it not work again.
So I try add a new Inbound Rules in `Windows Defender FGirewall with Advanced Security`,and didnot change any options but display name is `WSL my`,so I got this:

![Image](https://github.com/user-attachments/assets/026bfe3c-c621-4084-8baa-89097c92edaf)

It works,after I rebooted windows ,it works still.

Maybe I know why `WSL by interface` didnot work,maybe cause windows didnot start wsl interface when windows boot so this rules is not use for wsl interface ,But I donot know why `WSL my` working,they are same rules. 😕 Anyway ,seems it a why to fix this problem.
